### PR TITLE
Feature/us5398 prepopulatefields

### PR DESCRIPTION
**Pull request by lknair2015**

### CODE REVIEW

File: Gateway/crds-angular/Controllers/API/CampController.cs [hunk 1]
```diff
@@ -7,15 +7,18 @@
...
+using MinistryPlatform.Translation.Repositories.Interfaces;
 
 namespace crds_angular.Controllers.API
 {
     public class CampController: MPAuth
     {
         private readonly ICampService _campService;
-        public CampController(ICampService campService)
+        private readonly IContactRepository _contactService;
+        public CampController(ICampService campService, IContactRepository contactService)
```
Comment (silbermm) [inline Gateway/crds-angular/Controllers/API/CampController.cs]: we should not be injecting translation layer repositories directly in a controller. It should be abstracted out into a gateway layer service.

File: Gateway/crds-angular/Controllers/API/CampController.cs [hunk 2]
```diff
@@ -58,6 +61,34 @@ public IHttpActionResult GetCampEventDetails(int eventId)
...
+        [Route("api/camps/{eventid}/{contactid}")]
+        [AcceptVerbs("GET")]
+        public IHttpActionResult GetCamperInfo(int eventId, int contactId)
+        {
+            return Authorized(token =>
+            {
+                try
+                {
+                    var loggedInContact = _contactService.GetMyProfile(token);
+                    var family = _contactService.GetHouseholdFamilyMembers(loggedInContact.Household_ID);
```
Comment (silbermm) [inline Gateway/crds-angular/Controllers/API/CampController.cs]: this logic should live in a method in the service layer. We should be doing very little in the controller

File: Gateway/crds-angular/Services/CampService.cs [hunk 3]
```diff
@@ -138,5 +143,34 @@ public List<MyCampDTO> GetMyCampInfo(string token)
...
+
+        public CampReservationDTO GetCamperInfo(string token, int eventId, int contactId)
+        {
+            var camperContact = _contactService.GetContactById(contactId);
+            var participant = _participantRepository.GetParticipant(contactId);
+            var gradeGroupTypeId = _configurationWrapper.GetConfigIntValue("AgeorGradeGroupType");
+            var grageGoup = _groupService.GetGroupsByTypeForParticipant(token, participant.ParticipantId, gradeGroupTypeId);
+
+            string currentGrade = null;
+            foreach (var group in grageGoup)
```
Comment (silbermm) [inline Gateway/crds-angular/Services/CampService.cs]: instead of looping through the grade groups it would be best to just get the first one `gradeGroup.FirstOrDefault()`

File: Gateway/crds-angular/Services/CampService.cs [hunk 4]
```diff
@@ -138,5 +143,34 @@ public List<MyCampDTO> GetMyCampInfo(string token)
 
             return dashboardData;
         }
+
+        public CampReservationDTO GetCamperInfo(string token, int eventId, int contactId)
+        {
+            var camperContact = _contactService.GetContactById(contactId);
+            var participant = _participantRepository.GetParticipant(contactId);
+            var gradeGroupTypeId = _configurationWrapper.GetConfigIntValue("AgeorGradeGroupType");
+            var grageGoup = _groupService.GetGroupsByTypeForParticipant(token, participant.ParticipantId, gradeGroupTypeId);
```
Comment (silbermm) [inline Gateway/crds-angular/Services/CampService.cs]: whats a `grageGoup` :wink:

File: crossroads.net/app/camps/camp.controller.js [hunk 5]
```diff
@@ -2,7 +2,6 @@
 class CampController {
   constructor(CampsService) {
     this.viewReady = false;
-    this.campsService = CampsService;
```
Comment (silbermm) [inline crossroads.net/app/camps/camp.controller.js]: why remove the service from the controller? we still use it (look at the `$onInit()` method)

File: crossroads.net/app/camps/camper_info/camper_info_form.service.js [hunk 6]
```diff
@@ -35,7 +36,7 @@ class CamperInfoForm {
           {
             className: 'form-group col-xs-6',
             key: 'firstName',
-            type: 'crdsInput',
+            type: 'crdsInput',
```
Comment (silbermm) [inline crossroads.net/app/camps/camper_info/camper_info_form.service.js]: whitespace?

File: crossroads.net/app/camps/camps.routes.js [hunk 7]
```diff
@@ -18,27 +18,56 @@ export default function CampRoutes($stateProvider) {
...
-        getCampInfo: getCampInfo,
+        getCamperInfo: getCamperInfo,
         $stateParams: '$stateParams'
-      }
-    });
+      }       
+    });    
+}
+
+function getCamperInfo(CampsService, $stateParams) {
```
Comment (silbermm) [inline crossroads.net/app/camps/camps.routes.js]: functions should be at defined before they are used. this should be moved to the top of the file

File: crossroads.net/app/camps/camps.routes.js [hunk 8]
```diff
@@ -18,27 +18,56 @@ export default function CampRoutes($stateProvider) {
...
+  let camperId = $stateParams.camperId;
+  let campId = $stateParams.campId;
+  return CampsService.getCamperInfo(campId, camperId);
 }
 
+
 function getCampInfo(CampsService, $stateParams) {
```
Comment (silbermm) [inline crossroads.net/app/camps/camps.routes.js]: this function has been moved in another pull request, no need to move this to the top of the file.

File: crossroads.net/app/camps/camps.routes.js [hunk 9]
```diff
@@ -18,27 +18,56 @@ export default function CampRoutes($stateProvider) {
...
-          title: 'Camp Signup',
+          title: 'Camper Signup',
           description: 'Join us for camp!'
         }
       },
       resolve: {
         loggedin: crds_utilities.checkLoggedin,
         campsService: 'CampsService',
-        getCampInfo: getCampInfo,
+        getCamperInfo: getCamperInfo,
```
Comment (silbermm) [inline crossroads.net/app/camps/camps.routes.js]: this can just be `getCamperInfo`. If the key and the value are the same name, you don't have to specify both (yea ES2015!)

File: Gateway/crds-angular/Models/Crossroads/Camp/CampReservationDTO.cs [hunk 10]
```diff
@@ -18,7 +21,7 @@ public class CampReservationDTO
         public string PreferredName { get; set; }
 
         [JsonProperty(PropertyName = "birthDate")]
-        public DateTime BirthDate { get; set; }
+        public string BirthDate { get; set; }
```
Comment (silbermm) [inline Gateway/crds-angular/Models/Crossroads/Camp/CampReservationDTO.cs]: why is this changed to a string type?

File: Gateway/crds-angular/Services/CampService.cs [hunk 11]
```diff
@@ -58,16 +62,18 @@ public CampDTO GetCampEventDetails(int eventId)
         public void SaveCampReservation(CampReservationDTO campReservation, int eventId, string token)
         {
             var parentContact = _contactService.GetMyProfile(token);
+            var congregationId = campReservation.CrossroadsSite;
+            var contactId = campReservation.ContactId;
```
Comment (silbermm) [inline Gateway/crds-angular/Services/CampService.cs]: it seems kinda odd that you set these to variables. Why not just reference them as is when needed?

File: Gateway/crds-angular/Services/CampService.cs [hunk 12]
```diff
@@ -58,16 +62,18 @@ public CampDTO GetCampEventDetails(int eventId)
...
             var displayName = campReservation.PreferredName;
             var nickName = displayName ?? campReservation.FirstName;
-            displayName = displayName != null ? campReservation.LastName + ", " + campReservation.PreferredName : campReservation.LastName + ", " + campReservation.FirstName;
+            displayName = displayName != null ? campReservation.PreferredName : campReservation.LastName + ", " + campReservation.FirstName;
```
Comment (silbermm) [inline Gateway/crds-angular/Services/CampService.cs]: why initialize `displayName` and then check it for null? Would't this be better like:
`var displayName = campReservation.PreferredName != null ? campReservation.PreferredName : campReservation.LastName + ", " + campReservation.FirstName;`
Then remove line 67.

File: Gateway/crds-angular/Services/CampService.cs [hunk 13]
```diff
@@ -58,16 +62,18 @@ public CampDTO GetCampEventDetails(int eventId)
...
             
             var minorContact = new MpContact
             {
                 FirstName = campReservation.FirstName,
                 LastName = campReservation.LastName,
                 MiddleName = campReservation.MiddleName,
-                BirthDate = campReservation.BirthDate,
+                BirthDate = Convert.ToDateTime(campReservation.BirthDate),
```
Comment (silbermm) [inline Gateway/crds-angular/Services/CampService.cs]: if we accept the param as a DateTime, then we won't have to convert

File: Gateway/MinistryPlatform.Translation/Repositories/CampRepository.cs [hunk 14]
```diff
@@ -33,6 +46,44 @@ public MpCamp GetCampEventDetails(int eventId)
             }            
             throw new Exception("No Camp found");
         }
-        
+
+        public Dictionary<string, object> SaveorUpdateCamperContactDetails(MpContact minorContact, int? contactId, int congregationId, string token)
```
Comment (silbermm) [inline Gateway/MinistryPlatform.Translation/Repositories/CampRepository.cs]: there should be some tests for this

File: Gateway/crds-angular/Services/CampService.cs [hunk 15]
```diff
@@ -58,16 +62,18 @@ public CampDTO GetCampEventDetails(int eventId)
         public void SaveCampReservation(CampReservationDTO campReservation, int eventId, string token)
```
Comment (silbermm) [inline Gateway/crds-angular/Services/CampService.cs]: tests???

File: Gateway/crds-angular/Controllers/API/CampController.cs (hunk 1 above)
Comment (lknair2015) [inline Gateway/crds-angular/Controllers/API/CampController.cs]: I updated it

File: Gateway/crds-angular/Controllers/API/CampController.cs (hunk 2 above)
Comment (lknair2015) [inline Gateway/crds-angular/Controllers/API/CampController.cs]: moved it to camp service

File: Gateway/crds-angular/Services/CampService.cs (hunk 13 above)
Comment (lknair2015) [inline Gateway/crds-angular/Services/CampService.cs]: I didnt want to update MpContact as it is used many other places

File: Gateway/crds-angular/Services/CampService.cs (hunk 4 above)
Comment (lknair2015) [inline Gateway/crds-angular/Services/CampService.cs]: updated  the typo ... hmmm....

File: Gateway/crds-angular/Services/CampService.cs (hunk 3 above)
Comment (lknair2015) [inline Gateway/crds-angular/Services/CampService.cs]: changed as per comment

File: crossroads.net/app/camps/camp.controller.js (hunk 5 above)
Comment (lknair2015) [inline crossroads.net/app/camps/camp.controller.js]: updated it

File: crossroads.net/app/camps/camper_info/camper_info_form.service.js (hunk 6 above)
Comment (lknair2015) [inline crossroads.net/app/camps/camper_info/camper_info_form.service.js]: removed it

File: crossroads.net/app/camps/camps.routes.js (hunk 9 above)
Comment (lknair2015) [inline crossroads.net/app/camps/camps.routes.js]: removed it

File: crossroads.net/app/camps/camps.routes.js (hunk 7 above)
Comment (lknair2015) [inline crossroads.net/app/camps/camps.routes.js]: moved them

File: crossroads.net/app/camps/camps.routes.js (hunk 8 above)
Comment (lknair2015) [inline crossroads.net/app/camps/camps.routes.js]: moved it

File: Gateway/crds-angular/Services/CampService.cs [hunk 16]
```diff
@@ -93,7 +98,7 @@ public void SaveCampReservation(CampReservationDTO campReservation, int eventId,
             var formId = _configurationWrapper.GetConfigIntValue("SummerCampFormID");
             var formResponse = new MpFormResponse
             {
-                ContactId = contactId,
+                ContactId = Convert.ToInt32(contactId),
```
Comment (lknair2015) [inline Gateway/crds-angular/Services/CampService.cs]: yea I updated it

File: Gateway/crds-angular/Models/Crossroads/Camp/CampReservationDTO.cs (hunk 10 above)
Comment (lknair2015) [inline Gateway/crds-angular/Models/Crossroads/Camp/CampReservationDTO.cs]: it is giving me an error

File: Gateway/MinistryPlatform.Translation/Repositories/CampRepository.cs (hunk 14 above)
Comment (lknair2015) [inline Gateway/MinistryPlatform.Translation/Repositories/CampRepository.cs]: I added them

File: Gateway/crds-angular/Services/CampService.cs (hunk 15 above)
Comment (lknair2015) [inline Gateway/crds-angular/Services/CampService.cs]: added 3 tests to camp service
